### ISSUE TITLE
return_index option for torch.unique

### ISSUE THREAD

**Issue by tuananhle7** [pytorch/pytorch#36748]
## 🚀 Feature
`return_index` option for [torch.unique](https://pytorch.org/docs/stable/torch.html#torch.unique) which behaves like [numpy.unique](https://docs.scipy.org/doc/numpy/reference/generated/numpy.unique.html).

I have tensor `a = [10, 20, 10, 30]` and tensor `b = [100, 200, 100, 300]`.
I want to take the unique elements of `a` (`[10, 20, 30]`), but also get the corresponding elements of `b` (`[100, 200, 300]`). Having the above feature would allow me to use the return indices to index directly to `b` for this.

Is this on the roadmap? Or are there alternative ways of achieving this?

Thank you!

**Comment by mruberry** [comment]
Thanks for the suggestion, tuananhle7! We are always working on improving our NumPy compatibility, but a PR implementing this behavior would be cool, too ;)

**Comment by janblumenkamp** [comment]
Would love to see such a feature! In the meantime, the outdated [pytorch_unique package](https://github.com/rusty1s/pytorch_unique) demonstrates a way to achieve this:

> PyTorch 0.4.1 now supports unique both for CPU and GPU. Therefore, this package is no longer needed and will not be updated. In contrast to this package, PyTorch's version does not return an index array. However, you can easily generate it by using the following code:

```
import torch

unique, inverse = torch.unique(x, sorted=True, return_inverse=True)
perm = torch.arange(inverse.size(0), dtype=inverse.dtype, device=inverse.device)
inverse, perm = inverse.flip([0]), perm.flip([0])
perm = inverse.new_empty(unique.size(0)).scatter_(0, inverse, perm)
```
Not sure if that is the most efficient way though.

**Comment by tuananhle7** [comment]
Thanks @janblumenkamp this works well. It seems to work for a `dim` argument as well. I've written a function based on your snippet in case it's useful to anyone reading this thread.

```
def unique(x, dim=None):
    """Unique elements of x and indices of those unique elements
    https://github.com/pytorch/pytorch/issues/36748#issuecomment-619514810

    e.g.

    unique(tensor([
        [1, 2, 3],
        [1, 2, 4],
        [1, 2, 3],
        [1, 2, 5]
    ]), dim=0)
    => (tensor([[1, 2, 3],
                [1, 2, 4],
                [1, 2, 5]]),
        tensor([0, 1, 3]))
    """
    unique, inverse = torch.unique(
        x, sorted=True, return_inverse=True, dim=dim)
    perm = torch.arange(inverse.size(0), dtype=inverse.dtype,
                        device=inverse.device)
    inverse, perm = inverse.flip([0]), perm.flip([0])
    return unique, inverse.new_empty(unique.size(0)).scatter_(0, inverse, perm)
```

**Comment by arbabu123** [comment]
it would be really useful for one of the usecase that im working on. Please consider prioritizing it. Thank you! cc @mruberry

**Comment by ngimel** [comment]
Related: #16330

**Comment by wpalfi** [comment]
@janblumenkamp @tuananhle7 
`scatter_()` documentation says 

> When indices are not unique, the behavior is non-deterministic (one of the values from src will be picked arbitrarily) 

https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html

**Comment by scaomath** [comment]
> @janblumenkamp @tuananhle7
> `scatter_()` documentation says
> 
> > When indices are not unique, the behavior is non-deterministic (one of the values from src will be picked arbitrarily)
> 
> https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html

I believe the "behavior" is referring to auto-differentiation. If just calling the routine above, the returned indices will be deterministic, as the following code snippet

```python
inverse, perm = inverse.flip([0]), perm.flip([0])
inverse.new_empty(unique.size(0)).scatter_(0, inverse, perm)
```

essentially does is:

```python
indices = inverse.new_empty(unique.size(0))
for i, ix in enumerate(inverse):
    indices[ix] = perm[i]
```

Therefore, if there are repeated indices in `inverse`, the second time the iteration gets a repeated one, the first one will get over-written. Now because both `inverse` and `perm` are flipped, you get a deterministic tensor which records of the indices of the first occurrence of `unique` tensor in the input tensor.

**Comment by jdebecdelievre** [comment]
Thank you very much for proposing the code snippet. If I am right, there is a typo and the returned index tensor should be created as `inverse.new_empty(unique.size(dim))` instead of `inverse.new_empty(unique.size(0))`.

Updated function:

`def unique(x, dim=None):
    """Unique elements of x and indices of those unique elements
    https://github.com/pytorch/pytorch/issues/36748#issuecomment-619514810

    e.g.

    unique(tensor([
        [1, 2, 3],
        [1, 2, 4],
        [1, 2, 3],
        [1, 2, 5]
    ]), dim=0)
    => (tensor([[1, 2, 3],
                [1, 2, 4],
                [1, 2, 5]]),
        tensor([0, 1, 3]))
    """
    unique, inverse = torch.unique(
        x, sorted=True, return_inverse=True, dim=dim)
    perm = torch.arange(inverse.size(0), dtype=inverse.dtype,
                        device=inverse.device)
    inverse, perm = inverse.flip([0]), perm.flip([0])
    return unique, inverse.new_empty(unique.size(dim)).scatter_(0, inverse, perm)`

**Comment by namespace-Pt** [comment]
Requesting the same feature.

**Comment by namespace-Pt** [comment]
Thank for the above answers. I modified several typos and I think anyone can just use this function:
```
def unique(x, dim=-1):
    unique, inverse = torch.unique(x, return_inverse=True, dim=dim)
    perm = torch.arange(inverse.size(dim), dtype=inverse.dtype, device=inverse.device)
    inverse, perm = inverse.flip([dim]), perm.flip([dim])
    return unique, inverse.new_empty(unique.size(dim)).scatter_(dim, inverse, perm)
```
If `x` is already sorted, use `torch.unique_consecutive` instead of `torch.unique`.

**Comment by rgommers** [comment]
gh-70920 is relevant here too, we should implement the separate functions which will add support for indices return value, rather than extend the current `torch.unique`.

**Comment by ehsansaleh** [comment]
Here's one way to obtain the first indices without relying on `scatter_` and its non-determinisim:

```python
def unique(x, dim=0):
    unique, inverse, counts = torch.unique(x, dim=dim, 
        sorted=True, return_inverse=True, return_counts=True)
    inv_sorted = inverse.argsort(stable=True)
    tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
    index = inv_sorted[tot_counts]
    _, index = index.sort()
    return unique, inverse, counts, index
```
The last output (i.e., the `index` variable) contains the first occurrence indices.

Hope this helps.

**Comment by hongsukchoi** [comment]
@ehsansaleh 
I greatly appreciate your solution. I was using the sparse convolution of Spconv2 and tried to eliminate the randomness of building sparse conv tensors, when there are duplicate voxel locations. [Related issue](https://github.com/traveller59/spconv/issues/159)

It seems Spconv2 is using _scatter function, which is non-deterministic, and different sparse conv tensors were made everytime.  Your solution works so good in my case. I am using torch 1.10, which does not support `stable=True` for argsort, so below is what I did. I referred to [here](https://github.com/pytorch/pytorch/issues/38681) for stable argsort. Just in case for others. It's a bit messy, but okay for evaluating my spconv network. It would be better to use `segment_csr', as addressed  [here](https://github.com/rusty1s/pytorch_scatter/issues/226), but I had multiple installation issues, and keeping it for future...

```
def stable_argsort(arr, dim=-1, descending=False):
        import numpy as np
        arr_np = arr.detach().cpu().numpy()
        if descending:
            indices = np.argsort(-arr_np, axis=dim, kind='stable')
        else:
            indices = np.argsort(arr_np, axis=dim, kind='stable')
        return torch.from_numpy(indices).long().to(arr.device)
        
def unique(x, dim=0):
        unique, inverse, counts = torch.unique(x, dim=dim, return_inverse=True, return_counts=True)
        inv_sorted = stable_argsort(inverse, dim=0, descending=False)
        tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
        index = inv_sorted[tot_counts]
        index = index.sort().values
        return unique, inverse, counts, index
        val, inv, cnt, idx = unique(world_3d_voxel_coord, dim=1)

# voxel_feat: (N, 3), voxel_coord: (1, N, 3), voxel_coord has duplicates
voxel_feat = voxel_feat[idx]
voxel_coord = voxel_coord[:, idx]

```

**Comment by ehsansaleh** [comment]
Here's one way you can avoid needing stable sort:

```python
def unique(x, dim=0):
    unique, inverse, counts = torch.unique(x, dim=dim, 
        sorted=True, return_inverse=True, return_counts=True)
    decimals = torch.arange(inverse.numel(), device=inverse.device) / n
    inv_sorted = (inverse+decimals).argsort()
    tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
    index = inv_sorted[tot_counts]
    index = index.sort().values
    return unique, inverse, counts, index
```

Hope this helps.

**Comment by hongsukchoi** [comment]
@ehsansaleh 
Awesome! You saved me so much time.

**Comment by simonseo** [comment]
> ```python
> def unique(x, dim=0):
>     unique, inverse, counts = torch.unique(x, dim=dim, 
>         sorted=True, return_inverse=True, return_counts=True)
>     decimals = torch.arange(inverse.numel(), device=inverse.device) / inverse.numel()
>     inv_sorted = (inverse+decimals).argsort()
>     tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
>     index = inv_sorted[tot_counts]
>     index = index.sort().values
>     return unique, inverse, counts, index
> ```

@ehsansaleh I'm comparing your output to that from `np.unique` but it looks like the output index is a bit off in some cases:

```python
a = np.array([1, 2, 6, 4, 2, 3, 2])

print(np.unique(a, return_index=True, return_inverse=True))
# (array([1, 2, 3, 4, 6]),
#  array([0, 1, 5, 3, 2]),
#  array([0, 1, 4, 3, 1, 2, 1]))

print(unique(torch.from_numpy(a)))
# (tensor([1, 2, 3, 4, 6]),
#  tensor([0, 1, 4, 3, 1, 2, 1]),
#  tensor([1, 3, 1, 1, 1]),
#  tensor([0, 1, 2, 3, 5])) # this is wrong
```

I don't have any suggestions yet on what's going wrong.

**Comment by scaomath** [comment]
@ehsansaleh 

The numpy's `index` output is "The indices of the first occurrences of the unique values in the original array". If one ought to match a torch routine with that of numpy. The index output should be sorted, since you want `unique == x[index]` is True.

**Comment by ehsansaleh** [comment]
I think sorting the index is an isolated use case, and can be applied outside the `unique` function effortlessly. Based on the discussions, syncing the `index` and `counts` variables seems a higher priority to maintain. I will include the final recipes here.

* For PyTorch Versions 1.13 and above (i.e., capable of stable-sorting), you can use the following:
```python
def unique(x, dim=0):
    unique, inverse, counts = torch.unique(x, dim=dim, 
        sorted=True, return_inverse=True, return_counts=True)
    inv_sorted = inverse.argsort(stable=True)
    tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
    index = inv_sorted[tot_counts]
    return unique, inverse, counts, index
```
* For PyTorch Versions 1.12 and below (i.e., incapable of stable-sorting), you can use the following:
```python
def unique(x, dim=0):
    unique, inverse, counts = torch.unique(x, dim=dim, 
        sorted=True, return_inverse=True, return_counts=True)
    decimals = torch.arange(inverse.numel(), device=inverse.device) / inverse.numel()
    inv_sorted = (inverse+decimals).argsort()
    tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
    index = inv_sorted[tot_counts]
    return unique, inverse, counts, index
```

Hope this helps us reach a conclusion.

@ngimel Please let me know if you would consider adding this to the next versions, so that I prepare a reasonable pull request for `torch.unique`. A more efficient implementation could remain in the works, while this temporarily helps the people in need.

**Comment by semaphore-egg** [comment]
> I think sorting the index is an isolated use case, and can be applied outside the `unique` function effortlessly. Based on the discussions, syncing the `index` and `counts` variables seems a higher priority to maintain. I will include the final recipes here.
> 
> * For PyTorch Versions 1.13 and above (i.e., capable of stable-sorting), you can use the following:
> 
> ```python
> def unique(x, dim=0):
>     unique, inverse, counts = torch.unique(x, dim=dim, 
>         sorted=True, return_inverse=True, return_counts=True)
>     inv_sorted = inverse.argsort(stable=True)
>     tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
>     index = inv_sorted[tot_counts]
>     return unique, inverse, counts, index
> ```
> 
> * For PyTorch Versions 1.12 and below (i.e., incapable of stable-sorting), you can use the following:
> 
> ```python
> def unique(x, dim=0):
>     unique, inverse, counts = torch.unique(x, dim=dim, 
>         sorted=True, return_inverse=True, return_counts=True)
>     decimals = torch.arange(inverse.numel(), device=inverse.device) / inverse.numel()
>     inv_sorted = (inverse+decimals).argsort()
>     tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
>     index = inv_sorted[tot_counts]
>     return unique, inverse, counts, index
> ```
> 
> Hope this helps us reach a conclusion.
> 
> @ngimel Please let me know if you would consider adding this to the next versions, so that I prepare a reasonable pull request for `torch.unique`. A more efficient implementation could remain in the works, while this temporarily helps the people in need.

Shall we specify `dim` in `arg_sort` or something?

**Comment by scaomath** [comment]
> Shall we specify `dim` in `arg_sort` or something?

```python
tensor.argsort(dim=-1)
```
or
```python
torch.argsort(tensor, dim=-1)
```

**Comment by wjaekim** [comment]
I needed the indices not only of unique slices but also of _last occurring_ unique slices. Even further, some use cases required such indices for unsorted, unique elements. In fact, unsorted outputs are not currently supported in PyTorch whenever dim is not None (even if unique indices are not requested). If you consider inverse indices, this problem gets more complicated.

So, I extended the function to cover all possible use cases. Of course, this should be regarded as a temporary patch until these are hopefully implemented as PyTorch's backend code. But I doubt it---some of these functionalities are still not adopted in Numpy.

Here is the function:

```python
def uniqueXT(x, sorted=True, return_index=False, return_inverse=False, return_counts=False,
             occur_last=False, dim=None):
    if return_index or (not sorted and dim is not None):
        unique, inverse, counts = torch.unique(x, sorted=True,
            return_inverse=True, return_counts=True, dim=dim)
        inv_sorted, inv_argsort = inverse.flatten().sort(stable=True)

        if occur_last and return_index:
            tot_counts = (inverse.numel() - 1 - 
                torch.cat((counts.new_zeros(1),
                counts.flip(dims=[0]).cumsum(dim=0)))[:-1].flip(dims=[0]))
        else:
            tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
        
        index = inv_argsort[tot_counts]
        
        if not sorted:
            index, idx_argsort = index.sort()
            unique = (unique[idx_argsort] if dim is None else
                torch.index_select(unique, dim, idx_argsort))
            if return_inverse:
                idx_tmp = idx_argsort.argsort()
                inverse.flatten().index_put_((inv_argsort,), idx_tmp[inv_sorted])
            if return_counts:
                counts = counts[idx_argsort]

        ret = (unique,)
        if return_index:
            ret += (index,)
        if return_inverse:
            ret += (inverse,)
        if return_counts:
            ret += (counts,)
        return ret if len(ret)>1 else ret[0]
    
    else:
        return torch.unique(x, sorted=sorted, return_inverse=return_inverse,
            return_counts=return_counts, dim=dim)
```

**Comment by CompRhys** [comment]
BUMP this would be very helpful if it could be added.

**Comment by chenzhekl** [comment]
+1 for this feature. Sometimes I would like to filter out unique elements based on some transformation of the input. Returning indices into the input tensor makes this easy to implement.
